src/perl/Permabit/BlockDevice/VDO/LVMVDO/ThinPool.pm

```diff
@@ -0,0 +1,176 @@
+
+  my $newSize = $self->{volumeGroup}->alignToExtent($logicalSize);
+  $self->{volumeGroup}->resizeLogicalVolume($name, $newSize);
+  $self->{logicalSize} = $newSize;
+}
+
+########################################################################
+# @inherit
+##
+sub growPhysical {
+  my ($self, $physicalSize) = assertNumArgs(2, @_);
+
+  # resize the vdo pool
+  my $name = "$self->{deviceName}pool";
+
+  my $newSize = $self->{metadataSize} + $physicalSize;
+  $newSize = $self->{volumeGroup}->alignToExtent($newSize);
+  $self->{volumeGroup}->resizeLogicalVolume($name, $newSize);
+  $self->{physicalSize} = $newSize - $self->{metadataSize};
+}
+
+########################################################################
+# Rename a VDO device. Currently LVM managed devices are the only
+# ones that support rename.
+#
+# @param newName The new name for the VDO device.
+##
+sub renameVDO {
+  my ($self, $newName) = assertNumArgs(2, @_);
+  my $name = "$self->{deviceName}";
+  $self->{volumeGroup}->renameLogicalVolume($name, $newName);
+  $self->setDeviceName($newName);
+}
+
+########################################################################
+# @inherit
+##
+sub disableWritableStorage {
+  my ($self) = assertNumArgs(1, @_);
+  my $deviceName = "$self->{deviceName}pool_vdata";
+  my $vgName   = $self->{volumeGroup}->getName();
+  my $fullName = "$vgName-$deviceName";
+
+  $self->runOnHost("sudo dmsetup remove $fullName");
+}
+
+########################################################################
+# @inherit
+##
+sub enableWritableStorage {
+  my ($self) = assertNumArgs(1, @_);
+  my $deviceName = "$self->{deviceName}pool_vdata";
+
+  $self->{volumeGroup}->enableLogicalVolume($deviceName);
+
+  my $vgName   = $self->{volumeGroup}->getName();
+  my $fullName = "$vgName-$deviceName";
+  my $table = $self->runOnHost("sudo dmsetup table $fullName");
+
+  $self->{volumeGroup}->disableLogicalVolume($deviceName);
+
+  $self->runOnHost("sudo dmsetup create $fullName --table '$table'");
+}
+
+########################################################################
+# @inherit
+##
+sub disableReadableStorage {
+  my ($self) = assertNumArgs(1, @_);
+  $self->{volumeGroup}->disableLogicalVolume("$self->{deviceName}_vpool0_vdata");
+}
+
+########################################################################
+# @inherit
+##
+sub enableReadableStorage {
+  my ($self) = assertNumArgs(1, @_);
+  $self->{volumeGroup}->enableLogicalVolume("$self->{deviceName}_vpool0_vdata");
+}
+
+########################################################################
+# @inherit
+##
+sub getVDOStoragePath {
+  my ($self) = assertNumArgs(1, @_);
+  if (defined($self->{volumeGroup})) {
+    my $vgName      = $self->{volumeGroup}->getName();
+    my $storageName = "$vgName-$self->{deviceName}_vpool0_vdata";
+    my $path        = makeFullPath($self->{deviceRootDir}, $storageName);
+
+    return $self->_resolveSymlink($path);
+  }
+  return $self->getStoragePath();
+}
+
```

Comment:
These next three methods, (getVDOStoragePath(), setDeviceName() and setSymbolicPath()) look identical betweeo the two implementations except for the $storageName. It seems like it woudl be less repetitive to create a helper method to get the proper storage pool name, and then the rest of the functions could stay in the superclass.